File: python/trello/trello.py
#-*- coding:utf-8 -*-

class Trello:
  def __init__(self, configFileName):
    self.configFileName = configFileName

  def auth(self):
    from oauth import Oauth
    oauth = Oauth(self.configFileName)
    try:
      oauth.oauth()
    except:
      print('用户信息验证失败')

  def getConfig(self):
    from config import Config
    return Config(self.configFileName)

  def getData(self, api, queryString = ''):
    def printErrorMessage():
      from os import path
      print(
        '数据获取失败：网络异常或相关资源不存在。\n' +
        '如果网络正常且存在相关资源请执行%s auth进行用户信息验证'
        % path.abspath(__file__ + '/../../trello_client')
    )

    accessKeys = self._getAccessKey()
    if not accessKeys['key'] or not accessKeys['token']:
      printErrorMessage()
      return None

    from urllib2 import urlopen
    try:
      data = urlopen(
        '%s?key=%s&token=%s%s' %
        (api, accessKeys['key'], accessKeys['token'], queryString)
      ).read()

      import json
      return json.loads(data)
    except:
      printErrorMessage()
      return None

  def _getAccessKey(self):
    from os import path
    tokenFile = path.abspath(self.configFileName + '/../token')
    if not path.exists(tokenFile):
      return dict(key = '', token = '')

    with open(path.abspath(tokenFile), 'r') as fp:
      data = fp.readlines()
      if len(data) != 2:
        return dict(key = '', token = '')
      return dict(key = data[0].strip(), token = data[1].strip())

File: node/trello/trello.js
var Trello = {
  init: function(configFileName)
  {
    Trello.configFileName = configFileName;
  },

  auth: function()
  {
    try {
      require('./oauth').oauth(Trello.configFileName);
    } catch(e) {
      console.log('用户信息验证失败');
    }
  },

  getConfig: function()
  {
    var config = require('./config');
    config.init(Trello.configFileName);
    return config;
  },

  getData: function(api, queryString, callback)
  {
    function printErrorMessage()
    {
      console.log(
        '数据获取失败：网络异常或相关资源不存在。\n' +
        '如果网络正常且存在相关资源请执行%s auth进行用户信息验证',
        require('path').dirname(__dirname) + '/trello_client'
      );
    }

    var accessKeys = Trello._getAccessKey();
    if (!accessKeys['key'] || !accessKeys['token']) {
      return printErrorMessage();
    }

    queryString = queryString ? queryString : '';
    api += '?key=' + accessKeys['key'] + '&token=' + accessKeys['token'] + queryString;
    require('https').get(api, function(response) {
      response.on('data', function(data) {
        try {
          callback(JSON.parse(data.toString()));
        } catch(e) {
          printErrorMessage();
        }
      });
    }).on('error', printErrorMessage);
  },

  _getAccessKey: function()
  {
    try {
      var data = require('fs').readFileSync(
        require('path').dirname(Trello.configFileName) + '/token'
      ).toString().split('\n');
      if (data.length == 2) {
        return {key: data[0], token: data[1]};
      }
      return {key: null, token: null};
    } catch(e) {
      return {key: null, token: null};
    }
  }
};

module.exports = Trello;
